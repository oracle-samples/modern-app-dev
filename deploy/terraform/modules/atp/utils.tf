
resource "random_string" "autonomous_database_admin_password" {
  length           = 16
  special          = true
  min_upper        = 3
  min_lower        = 3
  min_numeric      = 3
  min_special      = 3
  override_special = "{}#^*<>[]%~"
}

resource "random_string" "autonomous_database_wallet_password" {
  length           = 16
  special          = true
  min_upper        = 3
  min_lower        = 3
  min_numeric      = 3
  min_special      = 3
  override_special = "{}#^*<>[]%~"
}