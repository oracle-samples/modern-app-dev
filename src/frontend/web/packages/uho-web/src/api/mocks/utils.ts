/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */

type Options = {
  delay?: number;
  error?: Error;
};
const defaultDelay = 0;

export function mock<T>(data: T, options?: Options): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    setTimeout(() => {
      options?.error ? reject(options.error) : resolve(data);
    }, options?.delay || defaultDelay);
  });
}
export function reject<T>(options?: Options): Promise<T> {
  return new Promise<T>((_, reject) => {
    setTimeout(() => {
      reject(options?.error);
    }, options?.delay || defaultDelay);
  });
}

export function mockDate() {
  return new Date(1654733487565);
}
