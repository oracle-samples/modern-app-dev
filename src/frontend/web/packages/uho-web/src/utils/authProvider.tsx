/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { h, ComponentChildren, createContext } from 'preact';
import { useCallback, useContext } from 'preact/hooks';
import { RouteProps } from 'react-router';
import { Navigate } from 'react-router-dom';
import { Loading } from 'components/loading';
import { useLocalStorage } from 'hooks/useLocalStorage';
import { frontendApi } from 'api';
import { User } from 'api/frontendApi';
import { useQuery } from '@tanstack/react-query';
import { warn } from 'ojs/ojlogger';

export type UserRole = 'PATIENT' | 'PROVIDER';

type Context = {
  user: User | null;
  signin: () => Promise<User | null>;
  signout: () => void;
};

function useProvideAuth(): Context {
  const [user, setUser] = useLocalStorage<User | null>('user');

  const signin = useCallback(async () => {
    let userInfo: User | null = null;
    try {
      userInfo = await frontendApi.getUserInformation();
    } catch {
      warn('No active user session found.');
    }
    setUser(userInfo);
    return userInfo;
  }, []);

  const signout = useCallback(async () => {
    setUser(null);
    return frontendApi.signout();
  }, []);

  return {
    user,
    signin,
    signout
  };
}

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
export const AuthContext = createContext<Context>(undefined!);

export function useAuth() {
  return useContext(AuthContext);
}

export function AuthProvider({ children }: { children: ComponentChildren }) {
  const auth = useProvideAuth();
  return <AuthContext.Provider value={auth}>{children}</AuthContext.Provider>;
}

export function RequireAuth({ children, role }: { children?: ComponentChildren; role: UserRole } & RouteProps) {
  const { signin } = useAuth();
  const { data, isLoading, isError } = useQuery<User | null, void>(['signin'], signin);

  if (isLoading) {
    return <Loading />;
  }
  if (!data || isError) {
    return <Navigate to="/" />;
  }
  if (data.role !== role) {
    return <Navigate to="/404" />;
  }
  return children;
}
