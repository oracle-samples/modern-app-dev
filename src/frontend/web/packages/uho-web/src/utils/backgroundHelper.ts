/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { ComponentProps } from 'preact/compat';
import { Avatar } from 'oj-c/avatar';

type AvatarProps = ComponentProps<typeof Avatar>;
const backgrounds: AvatarProps['background'][] = [
  'neutral',
  'orange',
  'green',
  'teal',
  'blue',
  'slate',
  'pink',
  'purple',
  'lilac',
  'gray'
];

// simple hash function to generate a number from 0 - brackets
function computeHash(key: string, brackets: number) {
  const hash = key.length + key.charCodeAt(0);
  return hash % brackets || 0;
}

export function getBackgroundByName(name: string) {
  return backgrounds[computeHash(name, backgrounds.length)];
}
