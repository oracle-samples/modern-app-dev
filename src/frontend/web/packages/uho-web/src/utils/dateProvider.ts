/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */

// extracted such that the method can be mocked in tests for consistent snapshot tests
export function now() {
  return new Date();
}
