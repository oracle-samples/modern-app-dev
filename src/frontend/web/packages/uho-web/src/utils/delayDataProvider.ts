/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import {
  ContainsKeysResults,
  DataProvider,
  FetchByKeysParameters,
  FetchByKeysResults,
  FetchByOffsetParameters,
  FetchByOffsetResults,
  FetchListParameters,
  FetchListResult
} from 'ojs/ojdataprovider';

const wrapPromise = <T>(promise: Promise<T>, delay: number) => {
  return new Promise<T>((resolve, reject) => {
    setTimeout(() => {
      promise.then(
        (result) => {
          resolve(result);
        },
        (reason) => {
          reject(reason);
        }
      );
    }, delay);
  });
};

class WrappingAsyncIterator<T> {
  constructor(public asyncIterator: AsyncIterator<T>, public delay: number) {}
  public next() {
    return wrapPromise(this.asyncIterator.next(), this.delay);
  }
}
export class DelayDataProvider<K, D> implements DataProvider<K, D> {
  constructor(public dataProvider: DataProvider<K, D>, public delay: number) {}

  addEventListener(eventType: string, listener: EventListener): void {
    return this.dataProvider.addEventListener(eventType, listener);
  }
  containsKeys(parameters: FetchByKeysParameters<K>): Promise<ContainsKeysResults<K>> {
    return wrapPromise(this.dataProvider.containsKeys(parameters), this.delay);
  }
  dispatchEvent(evt: Event): boolean {
    return this.dataProvider.dispatchEvent(evt);
  }
  fetchByKeys(parameters: FetchByKeysParameters<K>): Promise<FetchByKeysResults<K, D>> {
    return wrapPromise(this.dataProvider.fetchByKeys(parameters), this.delay);
  }
  fetchByOffset(parameters: FetchByOffsetParameters<D>): Promise<FetchByOffsetResults<K, D>> {
    return wrapPromise(this.dataProvider.fetchByOffset(parameters), this.delay);
  }
  fetchFirst(parameters?: FetchListParameters<D>): AsyncIterable<FetchListResult<K, D>> {
    const asyncIterable = this.dataProvider.fetchFirst(parameters);
    const asyncIterator = asyncIterable[Symbol.asyncIterator]();
    const wrappingAsyncIterator = new WrappingAsyncIterator(asyncIterator, this.delay);
    asyncIterable[Symbol.asyncIterator] = () => {
      return wrappingAsyncIterator;
    };
    return asyncIterable;
  }
  getCapability(capabilityName: string) {
    return this.dataProvider.getCapability(capabilityName);
  }
  getTotalSize(): Promise<number> {
    return this.dataProvider.getTotalSize();
  }
  isEmpty(): 'yes' | 'no' | 'unknown' {
    return this.dataProvider.isEmpty();
  }
  removeEventListener(eventType: string, listener: EventListener): void {
    return this.dataProvider.removeEventListener(eventType, listener);
  }
}
