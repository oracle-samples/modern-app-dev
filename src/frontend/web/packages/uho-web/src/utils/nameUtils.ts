/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { Patient } from '@uho/patient-api-client/dist/api-client';
import { Provider } from '@uho/provider-api-client/dist/api-client';

export function getInitialsFromPatient(patient?: Patient) {
  if (!patient || !patient.name) {
    return '';
  }
  return getInitialsFromName(...patient.name.split(' '));
}

export function getInitialsFromProvider(provider?: Provider) {
  if (!provider) {
    return '';
  }
  return getInitialsFromName(provider.firstName, provider.lastName);
}

export function getInitialsFromName(...names: (string | undefined)[]) {
  if (!names) {
    return '';
  }
  return names.map((part) => (part ? part.substring(0, 1).toUpperCase() : '')).join('');
}
