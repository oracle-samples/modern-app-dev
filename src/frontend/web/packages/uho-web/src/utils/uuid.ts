/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */

// based on https://stackoverflow.com/a/105078/1484047
export function generateGuid() {
  let result = '';
  let i;
  let j;
  for (j = 0; j < 32; j++) {
    if (j == 8 || j == 12 || j == 16 || j == 20) {
      result = `${result}-`;
    }
    i = Math.floor(Math.random() * 16)
      .toString(16)
      .toUpperCase();
    result = result + i;
  }
  return result;
}
