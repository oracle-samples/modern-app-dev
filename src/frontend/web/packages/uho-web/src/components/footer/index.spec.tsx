/*
 * Copyright (c) 2023 Oracle and/or its affiliates.
 * Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracl.com/licenses/upl/
 */
import { Footer, FooterElement } from '.';
import { h } from 'preact';
import { snapshot } from '../../../test/jetHelper';
import { MemoryRouter } from 'react-router-dom';
import { AuthProvider } from '../../utils/authProvider';

const elements: FooterElement[] = [
  {
    id: '1',
    path: 'foo',
    title: 'foo',
    iconClass: 'icon'
  }
];
snapshot(
  'Footer',
  <AuthProvider>
    <MemoryRouter>
      <Footer elements={elements} />
    </MemoryRouter>
  </AuthProvider>
);
