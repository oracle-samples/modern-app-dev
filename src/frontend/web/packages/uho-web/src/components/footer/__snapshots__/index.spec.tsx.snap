// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests for Footer Snapshot: Component - Footer 1`] = `
<DocumentFragment>
  <div
    data-oj-binding-provider="preact"
  >
    <div
      class="footer"
    >
      <oj-navigation-list
        class="oj-sm-condense oj-sm-margin-8x-horizontal oj-navigationlist oj-navigationlist-horizontal-bottom oj-navigationlist-expanded oj-navigationlist-horizontal oj-navigationlist-app-level oj-component-initnode oj-complete"
        role="tablist"
        tabindex="0"
      >
        <template
          slot="itemTemplate"
        />
        <div
          class="oj-navigationlist-listview-container"
          role="presentation"
        >
          <div
            class="oj-navigationlist-listview oj-component"
            role="presentation"
            style="max-width: none;"
          >
            <div
              class="oj-helper-detect-contraction"
            >
              <div
                style="width: 200%; height: 200%;"
              />
            </div>
            <div
              class="oj-helper-detect-expansion"
            >
              <div />
            </div>
            <div
              class="oj-helper-hidden-accessible"
              tabindex="-1"
            >
              Â 
            </div>
            <ul
              class="oj-navigationlist-element oj-navigationlist-has-icons"
              role="presentation"
            >
              <li
                class="oj-navigationlist-item-element oj-navigationlist-item oj-default"
                data-oj-vdom-template-root=""
                role="presentation"
              >
                <a
                  aria-selected="false"
                  class="oj-navigationlist-focused-element oj-navigationlist-item-content"
                  data-oj-tabmod="0"
                  href="#"
                  role="tab"
                  tabindex="-1"
                >
                  <span
                    class="oj-navigationlist-item-icon icon"
                  />
                  <span
                    class="oj-navigationlist-item-label"
                  >
                    foo
                  </span>
                </a>
              </li>
            </ul>
            <div
              class="oj-navigationlist-status-message oj-navigationlist-status"
              role="status"
              style="display: none;"
            >
              <div
                class="oj-icon oj-navigationlist-loading-icon"
              />
            </div>
            <div
              class="oj-helper-hidden-accessible"
              role="status"
              tabindex="-1"
            />
          </div>
        </div>
      </oj-navigation-list>
    </div>
  </div>
  <div
    id="__oj_layerhost_container"
  />
</DocumentFragment>
`;
